include/openthread/platform/toolchain.h

```diff
@@ -63,6 +63,22 @@
 extern "C" {
 #endif
 
+/**
+ * @def OT_MUST_USE_RESULT
+ *
+ * Compiler-specific indication that a class or enum must be used when it is
+ * the return value of a function.
+ *
+ * @note This is currently only available with clang (C++17 implements it
+ *       as attribute [[nodiscard]]).
+ *
+ */
+#if defined(__clang__)
+#define OT_MUST_USE_RESULT __attribute__((warn_unused_result))
+#else
+#define OT_MUST_USE_RESULT
+#endif
```

Comment:
```suggestion
#ifndef OT_MUST_USE_RESULT
#if defined(__clang__)
#define OT_MUST_USE_RESULT __attribute__((warn_unused_result))
#else
#define OT_MUST_USE_RESULT
#endif
#endif
```
A simple solution is allowing users to redefine OT_MUST_USE_RESULT so that they can turn off this feature?

Reply:
Is feature switching in public headers a good idea?

Reply:
@bukepo @abtink Any idea about this?

Reply:
@wgtdkp , yes we cannot use any config options in public headers (since we allow projects to define their own project header and since we don't own the file, we cannot make it part of public headers of OT). 

One way I can think of to allow user decide their own `MUST_USE_RESULT` is the following: (though I should say, it's complex and won't recommend it, just mentioning it as an option):
- We can add a core mirror of `otError` enumeration type (say `ot::Error`) which re-defiens the same enumeration values and use `OT_MUST_USE_RESULT` for `ot::Error` (similar to https://github.com/openthread/openthread/pull/4794 for `otDeviceRole` and `ot::Mle::DeviceRole`).  I think this can work (since the two types would be different and one can have `OT_MUST_USE_RESULT` and other don't).  This will  be big change and require a lot of `static_cast` in all the `{module}_api.cpp` and platform files to convert between `ot::Error` and `otError`.

Reply:
Thanks @abtink , it seems to be a big overhead. I come up with another solution: Instead of defining `OT_MUST_USE_RESULT` in public header, we can define it inside OT code and decorate `otError` with `OT_MUST_USE_RESULT` within only OT code. The price is, all OT core source file should always include a wrapper file (like `error_nodicard.hpp`) instead of `openthread/error.h` (This is not a big change, because we don't really explicitly includes `openthread/error.h` for each source file but inherit it) . The `error_nodiscard.hpp` file looks like this:

```c++
#if defined(__clang__)
#define OT_MUST_USE_RESULT __attribute__((warn_unused_result))
#else
#define OT_MUST_USE_RESULT
#endif

class OT_MUST_USE_RESULT otError;

#include <openthread/error.h>
```
In this way, we add `OT_MUST_USE_RESULT` attribute for only OT core code. Thoughts?

Reply:
I think this is the best solution so far.

Reply:
@abtink @jwhui thoughts?

Reply:
@wgtdkp , if it can be made to work reliably on all tool-chains, it is fine.

Note that OT public APIs are declared in C (not C++), so `otError` is a C `enum` declaration (not a `class` or `enum class`).  I think we'd need at least `C++11` to be able to use forward declaration of `enum` types.   OT currently requires/supports `C++03` and not yet on `C++11` (please see #https://github.com/openthread/openthread/issues/3880).
- Also If I remember correctly in C++11 we can only forward-declare an `enum` with a specified underlying (integral) type/size (or if it is a `enum class` and not `enum`).

Another point to conisder is if we can change the attribute on a forward declaration and not include the same attribute in the main definition. Would this work on all tool-chains?

Reply:
@abtink I will verify if `enum OT_MUST_USE_RESULT otError;` can work. To my understanding, the `warn_unused_result` attribute is specific to compiler and doesn't change generated code. So I think it is okay for application to not see this attribute with `otError`.

Reply:
@abtink , I verified that `enum OT_MUST_USE_RESULT otError;` works.